Mehrere Questions und Attempts ermöglichen

tumidi opened the pull request:
Issue: #165

- Es gibt jetzt IDs für Questions/Attempts, die in den URLs für's SDK Verwendung finden (bspw. `/question/abcd1234/attempt/efgh5678`. Der Speicherort für den State ändert sich entsprechend nach `PAKET/QUESTION_ID/ATTEMPT_ID/...`.

- Mir ist aufgefallen, dass überall eher der Name 'Options' verwendet wird. Jetzt, wo es mehrere Questions zu verwalten gibt, fand ich es passender und konsistenter 'Question' zu verwenden und habe daher einige Module und Symbole entsprechend umbenannt (im Back- sowie Frontend).

- Frontend: Es gab Umstrukturierungen bei den Pinia Stores. Ich habe die alten Stores für Attempts und Questions (bzw. Options früher) abgeschafft und durch Composables ersetzt. Vorher haben Stores (globaler State) Sinn gemacht, da es genau eine Question und einen Attempt gab.

- Frontend: Da wir jetzt Parameter in den API-URLs haben, verwende ich `unplugin-vue-router` für [typed routes](https://router.vuejs.org/guide/advanced/typed-routes.html) und bin bei der Gelegenheit auch direkt auf [file based routing](https://uvr.esm.is/guide/file-based-routing.html) gewechselt.

- Backend: Der Attempt-Controller wurde zunehmend unübersichtlicher und größer. Habe den einmal komplett refactored, eine Klasse `_AttemptRenderer` ausgegliedert, sowie die Datenklassen in ein eigenes Modul verlagert.

- Noch 2 offene Punkte:
  - Irgendwie ist das Verhalten beim Anlegen neuer Questions/Attempts nicht ganz konsistent:
    - Question wird bei Klick auf "New question" erst erzeugt wenn die Form abgeschickt wird (Geht auch nicht anders, da manche Questions eine ausgefüllte Form benötigen bzw. auch erst validiert werden müssen).
    - Bei Attempts dagegen ist es so, dass der Attempt sofort erzeugt wird, sobald auf "New attempt" geklickt wird.
    - Ich habe das für diesen PR erst mal ausgeklammert, aber evtl. wäre es sinnvoll, der Attempt-Seite auch eine Cancel- und Save-Funktionionalität zu geben (analog zu Questions). Ansonsten ist es etwas verwirrend/unerwartet, denke ich. Mich würde da auch eure Meinung zu interessieren.
  - Die `QuestionCard` ist aktuell noch rudimentär. Lars hatte da ja offenbar einen Teil der Question Options angeteasert (Mock-up), wobei ich mich frage, wie das für alle möglichen Options Forms gewährleistet werden kann. Aktuell wird als Platzhalter bloß die aktuelle Question-Form-Data angezeigt.

tumidi committed:
refactor(frontend): use `unplugin-vue-router` for typed file-system based routes

tumidi committed:
chore(frontend): update `@pinia/colada`

tumidi committed:
feat(webserver): add support for multiple questions and attempts

- Refactor `AttemptController`
  - Move rendering logic into `_AttemptRenderer`
- Add new API endpoints
- Rename `options` to `question`
- Define `StateManager` protocol and implement `FilesystemStateManager`

BREAKING CHANGE: Local state file format has changed

Ref: #165

tumidi committed:
feat(frontend): add support for multiple questions and attempts

- Transform global question/attempt stores into composables
- Rename `options` to `question`
- Introduce centralized query cache keys
- Use route names instead of hard-coded URL paths

Closes: #165

tumidi committed:
test: update tests to align with new API structure

janbritz commented on questionpy_sdk/webserver/state.py:
```diff
@@ -3,89 +3,230 @@
-    async def read_score(self) -> ScoreModel:
-        score_json = await self._read_state_file(StateFilename.SCORE)
+    async def read_question_states(self) -> dict[str, str]:
+        def _read_question_states() -> dict[str, str]:
+            try:
+                dir_iter = self._package_state_dir.iterdir()
+            except FileNotFoundError:
+                # Package dir might not have been created yet
+                return {}
+            else:
+                return {
+                    str(path.name): (path / self.StateFilename.QUESTION_STATE).read_text()
+                    for path in dir_iter
+                    if path.is_dir() and re.match(ID_RE, str(path.name))
+                }
```
1. `iterdir()` schmeißt interessanterweise, im Gegensatz zu Python 3.13, unter Python 3.12.2 keinen `FileNotFoundError`. Ich bin gestern durch Zufall darauf gestoßen und habe in [den Docs](https://docs.python.org/3/library/pathlib.html#pathlib.Path.iterdir) auch nichts dazu gefunden, dort steht nur:
`If the path is not a directory or otherwise inaccessible, OSError is raised.`
Ohne try-except-else ist es mMn auch etwas übersichtlicher.

3.  `path.name` ist schon ein `str`

```suggestion
            if not self._package_state_dir.exists():
                return {}

            return {
                path.name: (path / self.StateFilename.QUESTION_STATE).read_text()
                for path in self._package_state_dir.iterdir()
                if path.is_dir() and re.match(ID_RE, path.name)
            }
```

tumidi replied:
> 1. `iterdir()` schmeißt interessanterweise, im Gegensatz zu Python 3.13, unter Python 3.12.2 keinen `FileNotFoundError`.

Wie kommst du genau darauf? Beide Versionen verhalten sich gleich, was mich sonst auch eher wundern würde:


```py
# test.py
from pathlib import Path

try:
    Path("/does/not/exist").iterdir()
except OSError as e:
    print(type(e).__name__, e)
```

```sh
$ python --version
Python 3.12.2
$ python test.py
FileNotFoundError [Errno 2] No such file or directory: '/does/not/exist'
```

> Ohne try-except-else ist es mMn auch etwas übersichtlicher.

try/catch ist eigentlich die idiomatische Variante: *it is easier to ask forgiveness than permission*

Die Varianten sind übrigens auch nicht equivalent.

- Die if-Variante führt eine Race-Condition ein, die zu einer unbehandelten Exception führen kann:

```py
    if not self._package_state_dir.exists():
        return {}

    # zwischen den Calls verschwindet das Verzeichnis

    return {
        path.name: (path / self.StateFilename.QUESTION_STATE).read_text()
        for path in self._package_state_dir.iterdir()  # 💥 FileNotFoundError
        if path.is_dir() and re.match(ID_RE, path.name)
    }
```

- Die if-Variante braucht einen zusätzlichen System-Call.

EDIT:

So sieht der Code jetzt übrigens aus, der else-Zweig war ja eigentlich nicht nötig und ist vielleicht so auch ein bisschen lesbarer/übersichtlicher?

```py
            try:
                dir_iter = self._package_state_dir.iterdir()
            except FileNotFoundError:
                # Package dir might not have been created yet
                return {}

            return {
                path.name: (path / self.StateFilename.QUESTION_STATE).read_text()
                for path in dir_iter
                if path.is_dir() and re.match(ID_RE, path.name)
            }
```

tumidi replied:
> 2\. `path.name` ist schon ein `str`

:+1:

janbritz replied:
> Wie kommst du darauf?

Mir ist das aufgefallen, da ich den Ordner manuell gelöscht hatte und der Server dann beim Lesen der Question-State-Files einen `FileNotFoundError` geschmissen hat.
Ich habe es gerade nochmal mit deinem Beispiel, `pyenv`, sowie der aktuellsten Python 3.12 Version und der ersten Python 3.13 Version getestet:
```shell
$ pyenv shell 3.12.11

$ python --version
Python 3.12.11

$ python test.py

$ pyenv shell 3.13.0

$ python --version
Python 3.13.0

$ python test.py
FileNotFoundError [Errno 2] No such file or directory: '/does/not/exist'
```

Das scheint wohl ein Bug zu sein.

> try/catch ist eigentlich die idiomatische Variante: _it is easier to ask forgiveness than permission_

Da gebe ich dir grundsätzlich Recht.

> Die if-Variante führt eine Race-Condition ein, die zu einer unbehandelten Exception führen kann

Race-Conditions existieren auch im bestehenden Code und ich denke auch nicht, dass wir das ohne größeren Aufwand vermeiden können und sollten:

```python
            try:
                dir_iter = self._package_state_dir.iterdir()
            except FileNotFoundError:
                # Package dir might not have been created yet
                return {}

            # zwischen den Calls verschwindet das Verzeichnis

            return {
                path.name: (path / self.StateFilename.QUESTION_STATE).read_text()  # 💥 FileNotFoundError
                for path in dir_iter
                if path.is_dir() and re.match(ID_RE, path.name)
            }
```

> So sieht der Code jetzt übrigens aus, der else-Zweig war ja eigentlich nicht nötig und ist vielleicht so auch ein bisschen lesbarer/übersichtlicher?

Finde ich auch lesbarer. Ich weiß nicht wie verbreitet der `Path.iterdir()`-Bug ist, da wir Python 3.12 (EOL: `2028-10`) unterstützen, sollten wir vielleicht trotzdem eine andere Lösung wählen. Ich weiß auch nicht, ob ein potentieller Fix als Security-Fix durchgehen würde.

Ich habe auch gerade mit verschiedenen Online-Tools (z.B. [hier (Python 3.12.3)](https://www.onlinegdb.com/online_python_compiler)) und Python < 3.13 den Test-Code getestet und nie wurde ein Exception geschmissen.

janbritz replied:
Ich habe gerade einen offenen [Bug-Report](https://github.com/python/cpython/issues/129871) gefunden.

tumidi replied:
> Das scheint wohl ein Bug zu sein.

It's a feature! https://github.com/python/cpython/issues/78722

Wie dem auch sei, solange wir noch Python-Versionen <3.13 unterstützen, ist der pragmatische Umgang, den Iterator in eine `list()` zu wrappen, so dass er direkt ausgewertet wird.

Auf jeden Fall gut, dass dir das aufgefallen ist.

4511baa6f6f70eb4bf807ba9e775805c0886f003

> Finde ich auch lesbarer.

6dc55fa1d4ca599d049319bfd4e796daf8f2787e

janbritz commented on frontend/src/pages/index.vue:
```diff
@@ -0,0 +1,34 @@
+<route lang="json">
+{
+    "name": "index",
+    "meta": { "title": "Package preview" }
```
Um einheitlich mit den anderen `title`s zu bleiben:
```suggestion
    "meta": { "title": "Package Preview" }
```

tumidi replied:
cb1dc3893c3391ad3d3bb8ffe9b2f191500f669b

janbritz commented on frontend/src/components/attempt/AttemptCard.vue:
```diff
@@ -0,0 +1,107 @@
+            </BRow>
+            <BRow class="mb-3">
+                <BCol>Storage location</BCol>
+                <BCol class="fw-bold"></BCol>
```
Hier vielleicht noch ein `TODO` hinterlassen.

tumidi replied:
cb1dc3893c3391ad3d3bb8ffe9b2f191500f669b

janbritz commented on questionpy_sdk/webserver/state.py:
```diff
@@ -3,89 +3,230 @@
+            dir_iter = self._question_path(question_id).iterdir()
+        except FileNotFoundError:
+            # Question dir might not have been created yet
+            return attempts
```
s.o.

tumidi replied:
4511baa6f6f70eb4bf807ba9e775805c0886f003

janbritz reviewed (changes requested):
Hier sind noch ein paar Dinge, die mir beim Testen aufgefallen sind:
- Beim Erstellen einer Frage sind die `Create*`-Buttons disabled und werden erst enabled, sobald man etwas in der Form ändert. Wenn man z.B. eine Options-Form mit nur zwei Radio-Buttons hat, kann man die Frage nicht erstellen, wenn man das per Default ausgewählte Feld ausgewählt lassen möchte.
- Wenn man eine erstellte Frage bearbeiten möchte, wird die Options-Form nicht ganz richtig gerendert. Folgendes habe ich bemerkt:
  - Alle Felder sind leer, selbst wenn ein Value für ein Feld vorliegen sollte.
  - Eine `radio_group` welche eine `option(selected=True)` beinhaltet, wird ohne selektierte Option angezeigt.
  - Die `hide_if`-Condition wird nicht beachtet.
- Wenn man eine Frage per "Create"-Button erstellt und dann etwas ändert, sollte statt "Preview" -> "Save and preview" auf dem Preview-Button stehen.
- Wenn man in einem aktiven Attempt auf "Edit Question" und anschließend auf "Save and return" klickt, wird nur gespeichert und nicht auf die Attempt-Seite weitergeleitet. Gleiches gilt auch für "Preview" - ich überlege gerade ob wir hier wirklich beide Buttons bräuchten.
- Wenn man in einem Attempt ist und auf "Send and submit" klickt,
  - sollte der Button nicht nochmal geklickt werden dürfen.
  - wird "Current attempt" (unten auf der Seite) nicht geupdatet. Das ist z.B. auch der Fall, wenn man in "Current attempt" den Attempt löscht. Erst nach manuellem Neuladen der Seite, wir der aktuelle Status angezeigt.
- Sollten wir vielleicht auch den Attempt-Score löschen, mit vorheriger Warnung, sobald die dazugehörige Frage bearbeitet wurde? Es ist nicht unwahrscheinlich, dass der Score bei Änderung der Frage-Optionen, nicht mehr richtig ist, was dann auch verwirrend sein kann.

tumidi committed:
fix: remove unnecessary str() wrap around `path.name`

tumidi committed:
fix(frontend): fix a typo and add a TODO

tumidi committed:
fix: account for `iterdir()` behavior change in Python 3.13

In Python 3.13, `iterdir()` raises exceptions immediately instead of lazily.
This change forces early evaluation to ensure consistent behavior across
Python versions.

Ref: https://github.com/python/cpython/pull/107320

tumidi committed:
refactor: remove unnecessary else after except block

tumidi committed:
fix(frontend): prevent Create buttons from being deactivated when adding a question

tumidi committed:
fix(frontend): revise question form action buttons

Add missing `.value` access to `isClean` ref. It's now always possible to click "Save and preview"
when creating a question.

tumidi committed:
fix(frontend): init form data when queries already resolved

MHajoha commented:
Ich habe Jan gerade nichts hinzuzufügen ^^

> * Question wird bei Klick auf "New question" erst erzeugt wenn die Form abgeschickt wird (Geht auch nicht anders, da manche Questions eine ausgefüllte Form benötigen bzw. auch erst validiert werden müssen).
> * Bei Attempts dagegen ist es so, dass der Attempt sofort erzeugt wird, sobald auf "New attempt" geklickt wird.
> * Ich habe das für diesen PR erst mal ausgeklammert, aber evtl. wäre es sinnvoll, der Attempt-Seite auch eine Cancel- und Save-Funktionionalität zu geben (analog zu Questions). Ansonsten ist es etwas verwirrend/unerwartet, denke ich. Mich würde da auch eure Meinung zu interessieren.

Hierzu noch mein Senf: Moodle verwendet für Attempts auch andere Sprache als für Questions:
- Question: Add, Save, Edit, Delete
- Attempt: 
  - "Attempt Quiz" (also als Verb, bei uns dann "Attempt Question")
  - Submit / Finish
    - Es gibt auch noch den Lang-String "Save without submitting", der wird aber scheinbar nicht mehr verwendet
  - Continue
  - Delete (aus Trainer:innen-Perspektive)

Ich denke es wäre am besten, Moodle's Sprache (und die damit einhergehende Semantik) zu matchen, da in absehbarer Zukunft alle Menschen, die QPy verwenden werden, mit Moodle vertraut sind. Ich sehe auch das Argument, das, was technisch im Hintergrund passiert, genauer abzubilden, das verwirrt vermutlich noch mehr.

> Sollten wir vielleicht auch den Attempt-Score löschen, mit vorheriger Warnung, sobald die dazugehörige Frage bearbeitet wurde? Es ist nicht unwahrscheinlich, dass der Score bei Änderung der Frage-Optionen, nicht mehr richtig ist, was dann auch verwirrend sein kann.

Geiler, (aber auch aufwändiger, )wäre es, alle Attempts mit vorheriger Warnung zu rescoren. Die Option zum Löschen zu geben hate ich für einen einfacheren ersten Schritt.

MartinGauk commented:
> > Sollten wir vielleicht auch den Attempt-Score löschen, mit vorheriger Warnung, sobald die dazugehörige Frage bearbeitet wurde? Es ist nicht unwahrscheinlich, dass der Score bei Änderung der Frage-Optionen, nicht mehr richtig ist, was dann auch verwirrend sein kann.
> 
> Geiler, (aber auch aufwändiger, )wäre es, alle Attempts mit vorheriger Warnung zu rescoren. Die Option zum Löschen zu geben hate ich für einen einfacheren ersten Schritt.

An sich ist das auch ein Use-Case, den man ja im SDK eventuell testen will. Sobald Attempts in Moodle vorliegen, kann man die Datenstruktur der `scoring_state` und der Antworten nicht mehr verändern, man muss also abwärtskompatibel bleiben, sobald eine Paketversion einmal produktiv verwendet wurde.
Wobei man das natürlich am besten per automatisierten Tests sicherstellen sollte.

tumidi commented:
> - Wenn man eine erstellte Frage bearbeiten möchte, wird die Options-Form nicht ganz richtig gerendert. Folgendes habe ich bemerkt:
>   - Alle Felder sind leer, selbst wenn ein Value für ein Feld vorliegen sollte.
>   - Eine `radio_group` welche eine `option(selected=True)` beinhaltet, wird ohne selektierte Option angezeigt.
>   - Die `hide_if`-Condition wird nicht beachtet.

Da die Logik jetzt nicht mehr in einem globalen Store liegt, sondern in einem lokalen Composable, kann der Fall eintreten, dass die Queries resolven bevor die Komponente mountet. D.h. `watch`, was die `formData` initial zusammenbaut, läuft niemals. Der Fix ist `immediate: true` für `watch` zu setzen, was erzwingt, dass `watch` immer mind. einmal läuft.

59b58ce3e0c107cbf14b8fca4a409be5e9069dd2

> - Beim Erstellen einer Frage sind die `Create*`-Buttons disabled und werden erst enabled, sobald man etwas in der Form ändert. Wenn man z.B. eine Options-Form mit nur zwei Radio-Buttons hat, kann man die Frage nicht erstellen, wenn man das per Default ausgewählte Feld ausgewählt lassen möchte.

Ja, richtig. Im Falle einer neuen Frage sollte es immer möglich sein zu saven, denke ich.

0a02b717ad4a88fcf2cfeabe175573279e81c832

`saveAndPreview` verhält sich jetzt genauso, im Create-Fall ist es immer aktiviert, nur im Edit-Fall wird es deaktiviert falls es Validation-Fehler gibt.

e1a088a6c12a0873898bb97c3bcd8fc28a88ca6d

> - Wenn man eine Frage per "Create"-Button erstellt und dann etwas ändert, sollte statt "Preview" -> "Save and preview" auf dem Preview-Button stehen.

Hier fehlte das `.value` bei `isClean`.

e1a088a6c12a0873898bb97c3bcd8fc28a88ca6d

> - Wenn man in einem aktiven Attempt auf "Edit Question" und anschließend auf "Save and return" klickt, wird nur gespeichert und nicht auf die Attempt-Seite weitergeleitet.

Es wird gespeichert und auf die "Package Preview"-Seite navigiert, was das erwartete Verhalten ist.

Bei der Question-Form bedeutet "return" zur "Package Preview" zurückkehren (nicht Browser-History "Back"), während "preview" bedeutet: zur "Question Preview"-Seite navigieren.

Ich finde das auch verwirrend und wir sollten das alles vereinfachen mMn (siehe unten).

> Gleiches gilt auch für "Preview" -

Preview navigiert auf die (Question)-Preview-Seite, wie beabsichtigt.

> ich überlege gerade ob wir hier wirklich beide Buttons bräuchten.

Davon abgesehen, bin ich grundsätzlich aber auch nicht glücklich mit dem Design der Navigation und Action-Buttons allgemein. Wir sollten das überarbeiten, aber ich würde das gerne in einem separaten Issue tracken.

Ich hatte mich bei der Umsetzung halt erst mal an das Mock-up gehalten.

Auf manchen Seiten ist die Navigation oben als Button angeordnet, z.B. "Question Preview", oder auch "Attempt Preview". Manchmal ist sie in den Action Buttons kodiert, wie auf bei "Create/Edit Question" ("Save and return", "Save and Preview"). Bei "Attempt Preview" gibt es dann noch zusätzlich den "Edit question"-Link, der unter der Score erscheint.

"Return" ist missverständlich, wohin führt "return" überhaupt? Return (wie in Browser-History "back")? Das wäre wohl die intuitive Interpretation von "return". Aber wenn ich gerade von der Question-Preview-Seite komme würden dann beide Buttons ("Save and return" und "Save and preview") ja dasselbe machen, was auch weird ist. 

Das ganze ist auch in der Implementierung umständlich und fehleranfällig, da getrackt werden muss, welche Transitions gerade erlaubt sind und welche nicht, und wie diese dann korrekt bezeichnet werden...

> - Wenn man in einem Attempt ist und auf "Send and submit" klickt,
>  - sollte der Button nicht nochmal geklickt werden dürfen.

Ich würde das gerne erst mal ausklammern, zumal das ja vor diesem PR auch schon so war.

Die ganze UI-Logik sollte überarbeitet (vereinfacht) werden (siehe oben). Ich finde die Attempt-Seite auch recht verworren tbh. "Save and submit" bedeutet eigentlich "Save and score", aber ich kann auch erst auf "Save" und dann auf "Re-score" klicken stattdessen.

Der Attempt wird ja immer sofort gestartet, sobald ich einen Attempt aufrufe, d.h. der Status ist nur "Started" in dem Moment, wo ich einen nicht existierenden Attempt das erste Mal aufrufe. Refreshe ich die Seite oder navigiere weg, existiert der Attempt ab dem Moment und der Status springt plötzlich auf "In progress". Das bedeutet, vorher war der Restart-Button ausgegraut, plötzlich ist er es nicht mehr, obwohl ich keine Aktion ausgeführt habe (außer navigieren).

Ich könnte alternativ auch direkt einen zweiten Request nach dem Starten des Attempts machen, da sozusagen nach einmaligen Anzeigen des Attempts die Attemptdaten direkt "stale" sind (da der Attempt-Status ja auf "in progress" springt). Aber das kann es ja irgendwie nicht sein. Das ist es, was ich in der PR-Beschreibung unter "offene Punkte" kurz angerissen hatte.

>  - wird "Current attempt" (unten auf der Seite) nicht geupdatet. Das ist z.B. auch der Fall, wenn man in "Current attempt" den Attempt löscht. Erst nach manuellem Neuladen der Seite, wir der aktuelle Status angezeigt.

Ja, siehe oben. Die Attempt-Daten sind direkt "stale" nachdem ein Attempt gestartet wurde.

> - Sollten wir vielleicht auch den Attempt-Score löschen, mit vorheriger Warnung, sobald die dazugehörige Frage bearbeitet wurde? Es ist nicht unwahrscheinlich, dass der Score bei Änderung der Frage-Optionen, nicht mehr richtig ist, was dann auch verwirrend sein kann.

Ja, denke ich schon.

janbritz commented:
> Es wird gespeichert und auf die "Package Preview"-Seite navigiert, was das erwartete Verhalten ist.

> Preview navigiert auf die (Question)-Preview-Seite, wie beabsichtigt.

Beim ursprünglichen Testen, wurde tatsächlich gar nicht navigiert.

> Ich würde das gerne erst mal ausklammern, zumal das ja vor diesem PR auch schon so war.

:+1:
